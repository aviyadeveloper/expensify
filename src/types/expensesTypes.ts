import { v1 } from "uuid/interfaces";

type Expense = {
  id: v1;
  name: string;
  description: string;
  amount: number;
  createdAt: number;
};

export { Expense };
