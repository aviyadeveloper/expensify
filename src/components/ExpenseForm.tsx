import React from 'react';
import moment, { Moment } from 'moment';
import { SingleDatePicker } from 'react-dates';
import { Expense } from '../types/expensesTypes';
import { isNull } from 'util';

type ExpenseFormProps = {
  onSubmit?: (expense: Expense) => void;
  name?: string;
  description?: string;
  amount?: number;
  createdAt?: Moment;
  calendarFocused?: boolean | null;
  error?: string;
  editMode?: boolean;
};

interface ExpenseFormState {
  name: string;
  description: string;
  amount: string;
  createdAt: Moment;
  calendarFocused: boolean | null;
  error: string;
  editMode?: boolean;
}

export class ExpenseForm extends React.Component<
  ExpenseFormProps,
  ExpenseFormState
> {
  state: ExpenseFormState = {
    name: this.props.name || '',
    description: this.props.description || '',
    amount: (this.props.amount && (this.props.amount / 100).toFixed(2)) || '',
    createdAt:
      (this.props.createdAt && moment(this.props.createdAt)) || moment(),
    calendarFocused: false,
    error: '',
    editMode: this.props.editMode
  };

  onNameChange = (e: React.FormEvent<HTMLInputElement>) => {
    const name = e.currentTarget && e.currentTarget.value;
    this.setState({ name });
  };

  onAmountChange = (e: React.FormEvent<HTMLInputElement>) => {
    let amount = e.currentTarget && e.currentTarget.value;

    // Allow only cent based currency fit number format.
    (amount.match(/^[1-9]\d*(\.\d{0,2})?$/) || !amount) &&
      this.setState({ amount });
  };

  onDateChange = (createdAt: Moment | null) => {
    !isNull(createdAt) && this.setState({ createdAt });
  };

  onCalendarFocused = ({ focused }: { focused: boolean | null }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onDescriptionChange = (e: React.FormEvent<HTMLTextAreaElement>) => {
    let description = e.currentTarget && e.currentTarget.value;
    this.setState({ description });
  };

  onSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    if (!this.state.amount || !this.state.name) {
      this.setState(() => ({ error: 'Name and amount are required' }));
    } else {
      this.setState(() => ({ error: '' }));
      {
        this.props.onSubmit &&
          this.props.onSubmit({
            name: this.state.name,
            amount: parseFloat(this.state.amount) * 100,
            description: this.state.description,
            createdAt: this.state.createdAt
          });
      }
    }
  };

  render() {
    return (
      <div>
        <form onSubmit={this.onSubmit}>
          <input
            id="expense-form-name"
            type="text"
            placeholder="name"
            autoFocus
            value={this.state.name}
            onChange={this.onNameChange}
          />
          <input
            id="expense-form-amount"
            type="text"
            placeholder="amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            id="expense-form-date"
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={!!this.state.calendarFocused}
            onFocusChange={this.onCalendarFocused}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            id="expense-form-description"
            value={this.state.description}
            onChange={this.onDescriptionChange}
            placeholder="description"
          />
          {this.state.error && <p>{this.state.error}</p>}
          <button>{this.props.editMode ? 'Edit' : 'Add'}</button>
        </form>
      </div>
    );
  }
}
