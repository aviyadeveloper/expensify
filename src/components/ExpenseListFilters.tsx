import React from "react";
import { connect } from "react-redux";
import {
  setNameFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from "../actions/filtersActions";
import { FiltersSortBy } from "../types/filtersTypes";
import { DateRangePicker, FocusedInputShape } from "react-dates";
import { Moment } from "moment";

class ExpenseListFilters extends React.Component<any, any> {
  state = {
    focusedInput: null
  };

  nameFilter = React.createRef<HTMLInputElement>();
  sortByFilter = React.createRef<HTMLSelectElement>();

  onDatesChange = ({
    startDate,
    endDate
  }: {
    startDate: Moment | null;
    endDate: Moment | null;
  }) => {
    this.props.dispatch(setStartDate(startDate));
    this.props.dispatch(setEndDate(endDate));
  };

  onFocusChange = (focusedInput: "startDate" | "endDate" | null) => {
    this.setState({ focusedInput });
  };

  render() {
    return (
      <div>
        <input
          placeholder="filter by name"
          ref={this.nameFilter}
          value={this.props.filters.name}
          onChange={() => {
            this.props.dispatch(
              setNameFilter(
                (this.nameFilter.current && this.nameFilter.current.value) || ""
              )
            );
          }}
        />
        <select
          ref={this.sortByFilter}
          value={this.props.filters.sortBy}
          onChange={() => {
            if (this.sortByFilter.current) {
              const value: number = parseInt(this.sortByFilter.current.value);
              if (value === FiltersSortBy.Date) {
                return this.props.dispatch(sortByDate());
              }
              if (value === FiltersSortBy.Amount) {
                return this.props.dispatch(sortByAmount());
              }
            }
            //props.dispatch(sortByAmount());
          }}
        >
          <option value={FiltersSortBy.Date}>Date</option>
          <option value={FiltersSortBy.Amount}>Amount</option>
        </select>
        {this.props.filters.sortBy === FiltersSortBy.Date && (
          <DateRangePicker
            startDate={this.props.filters.startDate}
            startDateId={"sortByDate-filter-startDate"}
            endDate={this.props.filters.endDate}
            endDateId={"sortByDate-filter-endDate"}
            onDatesChange={this.onDatesChange}
            focusedInput={this.state.focusedInput}
            onFocusChange={this.onFocusChange}
            isOutsideRange={() => false}
            numberOfMonths={1}
            showClearDates={true}
          />
        )}
      </div>
    );
  }
}

const mapStateToProps = (state: any) => ({
  expenses: state.expenses,
  filters: state.filters
});

export default connect(mapStateToProps)(ExpenseListFilters);
