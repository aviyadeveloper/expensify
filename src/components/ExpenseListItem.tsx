import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { Expense } from "../types/expensesTypes";
import { removeExpense } from "../actions/expensesActions";

const ExpenseListItem = (props: Expense) => (
  <div>
    <Link to={`/edit/${props.id}`}>
      <h5>{props.name}</h5>
    </Link>
    <p>{props.description}</p>
    <p>
      <span>{props.amount}</span> / <span>{props.createdAt}</span>
    </p>
  </div>
);

export default connect()(ExpenseListItem);
