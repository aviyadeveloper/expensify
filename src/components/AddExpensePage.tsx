import React, { Dispatch } from 'react';
import { connect } from 'react-redux';
import { ExpenseForm } from './ExpenseForm';
import { Expense } from '../types/expensesTypes';
import { addExpense } from '../actions/expensesActions';
import { AnyAction } from 'redux';
import { History } from 'history';

type AddExpensePageProps = {
  onSubmit: (expense: Expense) => void;
  history: History;
};

export class AddExpensePage extends React.Component<AddExpensePageProps> {
  onSubmit = (expense: Expense): void => {
    this.props.onSubmit(expense);
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <h2>Expensify add expense page</h2>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch: Dispatch<AnyAction>) => ({
  onSubmit: (expense: Expense) => dispatch(addExpense(expense))
});

export default connect<any>(
  undefined,
  mapDispatchToProps
)(AddExpensePage);
