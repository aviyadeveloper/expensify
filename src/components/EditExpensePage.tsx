import React, { Dispatch } from 'react';
import { ExpenseForm } from './ExpenseForm';
import { Expense, expensesReducerState } from '../types/expensesTypes';
import { connect } from 'react-redux';
import { AnyAction } from 'redux';
import { editExpense, removeExpense } from '../actions/expensesActions';
import { History } from 'history';
import { match } from 'react-router';
import { storeStateType } from '../types/storeTypes';
import moment = require('moment');

export type EditExpensePageProps = {
  editExpense: (id: string, expense: Expense) => void;
  removeExpense: (id: string) => void;
  match: match<{ id: string | undefined }>;
  expense: Expense | undefined;
  history: History;
};

export class EditExpensePage extends React.Component<EditExpensePageProps> {
  onEditExpense = (expense: Expense) => {
    if (this.props.expense && typeof this.props.expense.id === 'string') {
      this.props.editExpense(this.props.expense.id, expense);
      this.props.history.push('/');
    }
  };

  onRemoveExpense = () => {
    if (this.props.expense && typeof this.props.expense.id === 'string') {
      this.props.removeExpense(this.props.expense.id);
      this.props.history.push('/');
    }
  };

  render() {
    return (
      <div>
        <h4>Editing expense</h4>
        <button id="remove-expense-button" onClick={this.onRemoveExpense}>
          Delete
        </button>
        <ExpenseForm
          onSubmit={this.onEditExpense}
          editMode={true}
          {...this.props.expense}
        />
      </div>
    );
  }
}

const mapStateToProps = (
  state: storeStateType,
  props: EditExpensePageProps
) => ({
  expense: state.expenses.find(
    (expense: Expense): boolean => expense.id === props.match.params.id
  )
});

const mapDispatchToProps = (dispatch: Dispatch<AnyAction>) => ({
  editExpense: (id: string, expense: Expense) => {
    dispatch(editExpense(id, expense));
  },
  removeExpense: (id: string) => {
    dispatch(removeExpense({ id }));
  }
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);
