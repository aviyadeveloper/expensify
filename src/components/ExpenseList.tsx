import React from "react";
import { connect } from "react-redux";
import { Expense } from "../types/expensesTypes";
import { FiltersReducerState } from "../types/filtersTypes";
import ExpenseListItem from "./ExpenseListItem";
import ExpenseListFilters from "./ExpenseListFilters";
import filterExpenses from "../selectors/expensesSelector";

type ExpenseListStateProps = {
  expenses: Expense[];
  filters: FiltersReducerState;
};

const ExpenseList = (props: ExpenseListStateProps) => (
  <div>
    <h1>Expense List:</h1>
    <ExpenseListFilters />
    <div>
      {props.expenses.map(e => (
        <ExpenseListItem key={e.id} {...e} />
      ))}
    </div>
  </div>
);

const mapStateToProps = (state: ExpenseListStateProps) => ({
  expenses: filterExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpenseList);
