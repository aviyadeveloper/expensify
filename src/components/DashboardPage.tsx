import React from 'react';
import { RouteProps } from 'react-router';
import ExpenseList from './ExpenseList';
import { filtersReducerDefaultState } from '../reducers/filtersReducer';

export const DashboardPage: React.SFC<RouteProps> = props => (
  <div>
    <ExpenseList filters={filtersReducerDefaultState} />
  </div>
);
