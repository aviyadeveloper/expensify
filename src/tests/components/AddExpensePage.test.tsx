import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { AddExpensePage } from '../../components/AddExpensePage';
import { createBrowserHistory, History } from 'history';
import { ExpenseForm } from '../../components/ExpenseForm';
import { expenses } from '../fixtures/expensesStateFixture';
import { Expense } from '../../types/expensesTypes';

let onSubmitSpy: (expense: Expense) => void,
  historySpy: History,
  wrapper: ShallowWrapper;

beforeEach(() => {
  onSubmitSpy = jest.fn();
  historySpy = createBrowserHistory();
  historySpy.push = jest.fn();
  wrapper = shallow(
    <AddExpensePage onSubmit={onSubmitSpy} history={historySpy} />
  );
});

test('should render AddExpensePage properly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () => {
  const onSubmit = wrapper.find(ExpenseForm).prop('onSubmit');
  onSubmit && onSubmit(expenses[0]);
  expect(onSubmitSpy).toHaveBeenLastCalledWith(expenses[0]);
  expect(historySpy.push).toHaveBeenLastCalledWith('/');
});
