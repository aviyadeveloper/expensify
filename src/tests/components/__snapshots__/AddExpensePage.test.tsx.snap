// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddExpensePage properly 1`] = `
<div>
  <h2>
    Expensify add expense page
  </h2>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
